lib/auth/apiserver.go

```diff
@@ -468,9 +468,9 @@ func (s *APIServer) generateUserCert(w http.ResponseWriter, r *http.Request, _ h
 	}
 	// SSH-to-HTTP gateway (tun server) sets HTTP basic auth to SSH cert principal
 	// This allows us to make sure that users can only request new certificates
-	// only for themselves
-	caller, _, _ := r.BasicAuth()
-	if req.User != caller {
+	// only for themselves, except admin users
+	caller, _, ok := r.BasicAuth()
+	if !ok || (req.User != caller && s.a.role != teleport.RoleAdmin) {
```

Comment:
Why not use `AuthWithRoles` directly instead of `a` for readability?
On the other note - if there's no BasicAuth in request, the error message will be somewhat dubious:
`User   cannot request a certificate for foo` - with nothing in place of the user name.